services/CertService.go

```diff
@@ -37,6 +47,81 @@ func bootstrap(tlsCertPath, privateKeyPath string) (*x509.Certificate, *rsa.Priv
 	}
 
 	cacheService := CacheService{}
-	cacheService.makeAndWatch()
+	cacheService.Watch()
 	return cer, key, nil
 }
+
+func createCertificateTemplate(serverName string, minutes time.Duration) *x509.Certificate {
+	cert := &x509.Certificate{
+		SerialNumber: big.NewInt(2019),
+		Subject: pkix.Name{
+			CommonName: serverName,
+		},
+		NotBefore:    time.Now(),
+		NotAfter:     time.Now().Add(time.Minute * minutes),
+		SubjectKeyId: []byte{1, 2, 3, 4, 6},
+		ExtKeyUsage:  []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:     x509.KeyUsageDigitalSignature,
+	}
+
+	if ip := net.ParseIP(serverName); ip != nil {
+		cert.IPAddresses = append(cert.IPAddresses, ip)
+	} else {
+		cert.DNSNames = append(cert.DNSNames, serverName)
+	}
+	return cert
+}
+
+func GenerateCertByName(serverName string) (*tls.Certificate, error) {
+	tx := apm.DefaultTracer.StartTransaction("ReturnCert("+serverName+")", "request")
+	defer tx.End()
+	ctx := apm.ContextWithTransaction(context.TODO(), tx)
+	tx.Context.SetLabel("inCache", "false")
+	cacheService := GetCacheService()
+	if serverCert, err := cacheService.FindCertinCache(serverName); err == nil {
+		tx.Context.SetLabel("inCache", "true")
+		return serverCert, nil
+	}
+	// set up our server certificate
```

Comment:
please extract a separate method here to reduce the complexity of the method GenerateCertByName and to remove the comment